### PR TITLE
[Frontend][Parse] Parse new named opaque return type syntax

### PR DESCRIPTION
**Pull request by willtunnels**
First commit in the process of implementing the proposal for named opaque return types and associated 'where' clause support. See discussion here: https://forums.swift.org/t/improving-the-ui-of-generics/22814.

### CODE REVIEW

**Comment by willtunnels** [comment]
@swift-ci please test

File: lib/Parse/ParsePattern.cpp [hunk 1]
```diff
@@ -864,17 +865,39 @@ Parser::parseFunctionSignature(Identifier SimpleName,
...
+
     ParserResult<TypeRepr> ResultType =
         parseDeclResultType(diag::expected_type_function_result);
     retType = ResultType.getPtrOrNull();
     Status |= ResultType;
     if (Status.isErrorOrHasCompletion())
       return Status;
 
+    if (GenericParams != nullptr) {
+      // The `Base` for our `OpaqueReturnParameterizedTypeRepr` should not be
```
Comment (slavapestov) [inline lib/Parse/ParsePattern.cpp:892]: The assert should go in the OpaqueReturnParameterizedTypeRepr constructor

File: test/type/opaque_experimental.swift [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// RUN: %target-swift-frontend -enable-experimental-opaque-return-types -disable-availability-checking -typecheck -verify %s
```
Comment (slavapestov) [inline test/type/opaque_experimental.swift:1]: Nitpick: you can use %target-typecheck-verify-swift to expand to %target-swift-frontend -typecheck -verify %s

File: test/type/opaque_experimental.swift [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+
+// Tests for experimental extensions to opaque return type support.
+
+func f0() -> <T> () { }
```
Comment (slavapestov) [inline test/type/opaque_experimental.swift:5]: Can you add tests with multiple generic parameters, and an inheritance clause on a generic parameter, as well as some invalid cases.

File: include/swift/AST/TypeRepr.h [hunk 4]
```diff
@@ -1101,6 +1101,42 @@ class OpaqueReturnTypeRepr : public TypeRepr {
...
 
+/// A TypeRepr for a type with a generic parameter list of named opaque return
+/// types.
+///
+/// This can occur only as the return type of a function declaration, to specify
+/// subtypes which should be abstracted from callers, given a set of generic
+/// constraints that the concrete types satisfy:
+///
+/// func foo() -> <T: Collection> T { return [1] }
+class OpaqueReturnParameterizedTypeRepr : public TypeRepr {
```
Comment (slavapestov) [inline include/swift/AST/TypeRepr.h:1112]: Nitpick: Maybe NamedOpaqueReturnTypeRepr? "Parametrized" is rather long

**Review by slavapestov — COMMENTED** [review]
Looks like a great start!

File: test/IDE/print_ast_named_opaque_return.swift [hunk 5]
```diff
@@ -0,0 +1,10 @@
+func f() -> <T> () {
+}
+func g() -> <T> Int {
```
Comment (slavapestov) [inline test/IDE/print_ast_named_opaque_return.swift:3]: Should try printing a generic parameter list with an inheritance clause as well

File: include/swift/AST/TypeRepr.h [hunk 6]
```diff
@@ -1101,6 +1101,44 @@ class OpaqueReturnTypeRepr : public TypeRepr {
   friend class TypeRepr;
 };
 
+/// A TypeRepr for a type with a generic parameter list of named opaque return
+/// types.
+///
+/// This can occur only as the return type of a function declaration, to specify
```
Comment (slavapestov) [inline include/swift/AST/TypeRepr.h:1107]: It can also occur as the return type of a computed property or subscript, no? (Those should be tested as well)

**Review by slavapestov — APPROVED** [review]
Looks ready to merge, except for the var/subscript decl tests I mentioned.

**Comment by willtunnels** [comment]
@swift-ci please smoke test

**Pull request merged**